cmd/helm/template.go

```diff
@@ -58,11 +64,51 @@ func newTemplateCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
 			if err != nil {
 				return err
 			}
-			fmt.Fprintln(out, strings.TrimSpace(rel.Manifest))
+
+			var manifests bytes.Buffer
+
+			fmt.Fprintln(&manifests, strings.TrimSpace(rel.Manifest))
 			if !client.DisableHooks {
 				for _, m := range rel.Hooks {
-					fmt.Fprintf(out, "---\n# Source: %s\n%s\n", m.Path, m.Manifest)
+					fmt.Fprintf(&manifests, "---\n# Source: %s\n%s\n", m.Path, m.Manifest)
+				}
+			}
+
+			// if we have a list of files to render, then check that each of the
+			// provided files exists in the chart.
+			if len(renderFiles) > 0 {
+				splitManifests := releaseutil.SplitManifests(manifests.String())
+				manifestNameRegex := regexp.MustCompile("# Source: [^/]+/(.+)")
+				var manifestsToRender []string
+				for _, f := range renderFiles {
+					missing := true
+					for _, manifest := range splitManifests {
+						submatch := manifestNameRegex.FindStringSubmatch(manifest)
+						if len(submatch) == 0 {
+							continue
+						}
+						manifestName := submatch[1]
+						// manifest.Name is rendered using linux-style filepath separators on Windows as
+						// well as macOS/linux.
+						manifestPathSplit := strings.Split(manifestName, "/")
+						manifestPath := filepath.Join(manifestPathSplit...)
+
+						// if the filepath provided matches a manifest path in the
+						// chart, render that manifest
+						if f == manifestPath {
+							manifestsToRender = append(manifestsToRender, manifest)
+							missing = false
+						}
+					}
+					if missing {
+						return fmt.Errorf("could not find template %s in chart", f)
+					}
+					for _, m := range manifestsToRender {
+						fmt.Fprintf(out, "---\n%s\n", m)
+					}
 				}
+			} else {
+				fmt.Fprintf(out, "%s", manifests.String())
```

Comment:
For anyone familiar with the helm-v2 code, this work is very similar to the implementation in v2. I just adapted it to work well with v3.

https://github.com/helm/helm/blob/66e0c0b44da1a7b604c592c87de88f545ce737bb/cmd/helm/template.go#L197-L231